src/base/bittorrent/bencoderesumedatastorage.cpp

```diff
@@ -134,42 +134,84 @@ QVector<BitTorrent::TorrentID> BitTorrent::BencodeResumeDataStorage::registeredT
     return m_registeredTorrents;
 }
 
-std::optional<BitTorrent::LoadTorrentParams> BitTorrent::BencodeResumeDataStorage::load(const TorrentID &id) const
+BitTorrent::LoadResumeDataResult BitTorrent::BencodeResumeDataStorage::load(const TorrentID &id) const
 {
     const QString idString = id.toString();
     const Path fastresumePath = m_resumeDataPath / Path(idString + u".fastresume");
     const Path torrentFilePath = m_resumeDataPath / Path(idString + u".torrent");
 
     QFile resumeDataFile {fastresumePath.data()};
     if (!resumeDataFile.open(QIODevice::ReadOnly))
-    {
-        LogMsg(tr("Cannot read file %1: %2").arg(fastresumePath.toString(), resumeDataFile.errorString()), Log::WARNING);
-        return std::nullopt;
-    }
+        return nonstd::make_unexpected(tr("Cannot read file %1: %2").arg(fastresumePath.toString(), resumeDataFile.errorString()));
 
     QFile metadataFile {torrentFilePath.data()};
     if (metadataFile.exists() && !metadataFile.open(QIODevice::ReadOnly))
-    {
-        LogMsg(tr("Cannot read file %1: %2").arg(torrentFilePath.toString(), metadataFile.errorString()), Log::WARNING);
-        return std::nullopt;
-    }
+        return nonstd::make_unexpected(tr("Cannot read file %1: %2").arg(torrentFilePath.toString(), metadataFile.errorString()));
 
     const QByteArray data = resumeDataFile.readAll();
     const QByteArray metadata = (metadataFile.isOpen() ? metadataFile.readAll() : "");
 
     return loadTorrentResumeData(data, metadata);
 }
 
-std::optional<BitTorrent::LoadTorrentParams> BitTorrent::BencodeResumeDataStorage::loadTorrentResumeData(
-        const QByteArray &data, const QByteArray &metadata) const
+void BitTorrent::BencodeResumeDataStorage::doLoadAll() const
+{
+    qDebug() << "Loading torrents count: " << m_registeredTorrents.size();
+
+    emit const_cast<BencodeResumeDataStorage *>(this)->loadStarted(m_registeredTorrents);
+
+    for (const TorrentID &torrentID : asConst(m_registeredTorrents))
+    {
+        nonstd::expected<LoadTorrentParams, QString> loadResumeDataResult = load(torrentID);
+        onResumeDataLoaded(torrentID, loadResumeDataResult);
+    }
+
+    emit const_cast<BencodeResumeDataStorage *>(this)->loadFinished();
+}
+
+void BitTorrent::BencodeResumeDataStorage::loadQueue(const Path &queueFilename)
+{
+    QFile queueFile {queueFilename.data()};
+    if (!queueFile.exists())
+        return;
+
+    if (!queueFile.open(QFile::ReadOnly))
+    {
+        LogMsg(tr("Couldn't load torrents queue: %1").arg(queueFile.errorString()), Log::WARNING);
+        return;
+    }
+
+    const QRegularExpression hashPattern {u"^([A-Fa-f0-9]{40})$"_qs};
+    QString line;
+    int start = 0;
+    while (!(line = QString::fromLatin1(queueFile.readLine().trimmed())).isEmpty())
```

Comment:
It would be nice to break it into smaller parts:
```c++
    while (true)
    {
        const auto line = QString::fromLatin1(queueFile.readLine().trimmed());
        if (line.isEmpty())
            break;
        
        // ...
    }
```

Reply:
When I studied programming, loops without a condition like `while (true)` were considered as something like `goto`.